MovingBlocks/FacadeServer
# Add secure websocket support

### Contains

Fix #11.

So far, I have looked on the internet to see how to add ssl support to jetty. The [Jetty website](http://www.eclipse.org/jetty/documentation/9.4.x/configuring-ssl.html) gives a good starting point. I created a keystore with the keytool like this:
![keystoreimg](https://user-images.githubusercontent.com/16342376/38273715-8abf661a-3741-11e8-9fdf-385b513231b3.PNG) (image is outdated, I made a different keystore with a few changes) as a way to test the server. Jetty supports SSL, so it isn't really that difficult to implement it.

For automatic certificate generation, it looks like we need a keystore. The Java [KeyStore class](https://docs.oracle.com/javase/8/docs/api/java/security/KeyStore.html) provides a way to load a keystore, although to automatically generate what we need for a self-signed certificate we need to do something else. At the moment the certificate I provided seems good enough.

### How to Test

You can test this by trying to login to wss://localhost:8443/ws in the [Facade Server Frontend](http://utility.terasology.org:8000/). Chrome appears to accept connections to the self-signed certificate on localhost, but for other browsers you will need to make an exception. You can also go to https://localhost:8443 in the browser and access the REST APIs over https, such as https://localhost:8443/api/resources/config/MOTD.

### Outstanding Before Merging
- [x] Generate self-signed certificate
- [x] Integrate new code into program better (allow specifying wss port, move some code into other parts if applicable)
- [x] Add a way for server owners to add their own certificate to the server
